src/ai/backend/manager/server.py

```diff
@@ -490,6 +491,113 @@ async def monitoring_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
+
+    from .models.session import SessionStatus
+
+    if TYPE_CHECKING:
+        from .models.utils import ExtendedAsyncSAEngine
+
+    async def _fetch_hanging_sessions(
+        db: ExtendedAsyncSAEngine,
+        status: SessionStatus,
+        threshold: timedelta,
+    ) -> tuple[SessionRow, ...]:
+        query = (
+            sa.select(SessionRow)
+            .where(SessionRow.status == status)
+            .where(
+                (
+                    datetime.now(tz=tzutc())
+                    - SessionRow.status_history[status.name].astext.cast(
+                        sa.types.DateTime(timezone=True)
+                    )
+                )
+                > threshold
+            )
+            .options(
+                noload("*"),
+                load_only(SessionRow.id, SessionRow.name, SessionRow.status, SessionRow.access_key),
+            )
+        )
+        async with db.begin_readonly() as conn:
+            result = await conn.execute(query)
+            return result.fetchall()
+
+    async def _force_terminate_hanging_sessions(
+        status: SessionStatus,
+        threshold: timedelta,
+        reason: KernelLifecycleEventReason = KernelLifecycleEventReason.HANGTIME_EXCEEDED,
+    ) -> None:
+        while True:
+            sessions = await _fetch_hanging_sessions(
+                root_ctx.db, status=status, threshold=threshold
+            )
+            log.debug(f"{len(sessions)} {status} sessions found.")
+
+            _ = await asyncio.gather(
+                *[
+                    asyncio.create_task(
+                        root_ctx.registry.destroy_session(session, forced=True, reason=reason),
+                    )
+                    for session in sessions
+                ]
+            )
+
+            await asyncio.sleep(threshold.seconds)
+
+    raw_session_config = await root_ctx.shared_config.etcd.get_prefix("config/session")
+    assert isinstance(raw_session_config, Mapping)
+
+    hang_tolerance_threshold_dict = raw_session_config.get("hang-tolerance-threshold", {})
+    assert isinstance(hang_tolerance_threshold_dict, Mapping)
+
+    session_force_termination_tasks = []
+    for status, threshold_fmt in hang_tolerance_threshold_dict.items():
+        assert isinstance(threshold_fmt, str)
+        try:
+            session_status = SessionStatus[status]
+        except KeyError:
+            continue
+
+        match threshold_fmt[-1]:
+            case "h":
+                threshold = timedelta(hours=int(threshold_fmt[:-1]))
+            case "m":
+                threshold = timedelta(minutes=int(threshold_fmt[:-1]))
+            case "s":
+                threshold = timedelta(seconds=int(threshold_fmt[:-1]))
+            case _:
+                log.error(
+                    f'Skip "{status}:{threshold_fmt}" as it is an invalid argument. Supported'
+                    ' postfixes: "h", "m", "s"'
+                )
+                continue
```

Comment:
I think we could reuse `ai.backend.common.validators.TimeDuration`.
If we use it, it is no longer required to assert `isinstance(threshold_fmt, str)` as well.

Reply:
Resolved: 20d878b5c4f2eadb538eeef4f577a011f7fb5887, c666327407e85fc27dc5dabd6cf2e426b318af3d, 9604d3c9a12970745df5c558d9c05064a4541eb2